### PR TITLE
space: more type annotations

### PR DESCRIPTION
**Pull request by Corvince**
Even more type annotations for space.py (after #779 and #796)

This PR makes mypy happy for grid, multigrid and hexgrid, except for some cases where the subclasses override the base class types (which is not solvable with type annotations, but would require changes to the logic itself).  

@rht since you did the previous type annotations, could you do a code review?

### CODE REVIEW

**Comment by rht** [comment]
Travis error: https://travis-ci.org/github/projectmesa/mesa/jobs/685832864#L363.

Also, it appears that mypy check hasn't been added to CI yet, might as well do it in this PR.

File: mesa/agent.py [hunk 1]
```diff
@@ -17,7 +19,7 @@ def __init__(self, unique_id: int, model: Model) -> None:
         """ Create a new agent. """
         self.unique_id = unique_id
         self.model = model
-        self.pos = None
+        self.pos: Optional[Union[Coordinate, FloatCoordinate, int]] = None
```
Comment (rht) [inline mesa/agent.py:22]: I think it would be cleaner to abstract out `Union[Coordinate, FloatCoordinate, int]`, so that when new coordinate type is added, agent.py will stay the same.

File: mesa/space.py [hunk 2]
```diff
@@ -349,6 +366,7 @@ def _place_agent(self, pos: Coordinate, agent: Agent) -> None:
     def remove_agent(self, agent: Agent) -> None:
         """ Remove the agent from the grid and set its pos variable to None. """
         pos = agent.pos
+        assert isinstance(pos, Tuple[int, int])
```
Comment (rht) [inline mesa/space.py:369]: Why is it hardcoded to be `Tuple[int, int]`, not `Coordinate`?

File: mesa/space.py [hunk 3]
```diff
@@ -856,7 +881,7 @@ def get_neighbors(self, node_id: int, include_center: bool = False) -> List[int]
 
     def move_agent(self, agent: Agent, node_id: int) -> None:
         """ Move an agent from its current node to a new node. """
-
+        assert isinstance(agent.pos, int)
```
Comment (rht) [inline mesa/space.py:884]: This assertion seems redundant if the user were to use mypy to make sure that `node_id` is `int`. If this assertion is added, then there are many places where similar assertions need to be added, and it might pollute the code.

File: mesa/agent.py (hunk 1 above)
Comment (Corvince) [reply mesa/agent.py:22]: Do you mean putting something like `Position = Union[Coordinate, FloatCoordinate, int]` into `space.py` and importing this in agent.py`?

File: mesa/space.py (hunk 2 above)
Comment (Corvince) [reply mesa/space.py:369]: Because otherwise mypy throws an error of "The type alias to Tuple is invalid in runtime context". However, at runtime `Tuple[int, int]` throws an "TypeError: Subscripted generics cannot be used with class and instance checks", so neither actually works. I will see what to do about this.

File: mesa/space.py (hunk 3 above)
Comment (Corvince) [reply mesa/space.py:884]: The assert is here to specifically make mypy happy. `agent.pos` might be of any type specified in the `agent.pos` type annotation and anything but int will throw an error in the `_remove_agent()` function (or, more specifically the call signature will not match)

**Comment by Corvince** [comment]
Thanks for the quick look through @rht! I see there are some issues to solve, I will ping you again once an update version is online. 

> Travis error: https://travis-ci.org/github/projectmesa/mesa/jobs/685832864#L363.
> 
> Also, it appears that mypy check hasn't been added to CI yet, might as well do it in this PR.

Maybe I understand something wrong about this, but unless we first resolve all mypy errors won't this just cause travis builds to always fail?

File: mesa/agent.py (hunk 1 above)
Comment (rht) [reply mesa/agent.py:24]: Yes.

**Comment by rht** [comment]
> Maybe I understand something wrong about this, but unless we first resolve all mypy errors won't this just cause travis builds to always fail?

Maybe without the `--strict` for now? Though I don't know how big the PR will be if all the `--strict` errors are addressed.

**Comment by Corvince** [comment]
@rht
I have a new version online. I primarily added a generic BaseGrid with Content as a type variable and changed the assertions to casts. Can you take another look?

Regarding travis CI: Somewhat related to this PR are still some type errors present for HexGrid, which redefines the call signature of several methods, which violates the Liskov Substitution Principle. I am not sure how to proceed with this. Aditionally there are some type errors for the ContinuousSpace, which I am currently not planning to address in this PR. Thus, even without `--strict` mypy will cry.

File: mesa/space.py [hunk 4]
```diff
@@ -118,29 +122,26 @@ def __init__(self, width: int, height: int, torus: bool) -> None:
...
+    def coord_iter(self) -> Iterator[Tuple[Content, int, int]]:
         """ An iterator that returns coordinates as well as cell contents. """
         for row in range(self.width):
             for col in range(self.height):
                 yield self.grid[row][col], row, col  # agent, x, y
 
-    def neighbor_iter(
-        self, pos: Coordinate, moore: bool = True
-    ) -> Iterator[GridContent]:
+    def neighbor_iter(self, pos: Coordinate, moore: bool = True) -> Iterator[Agent]:
```
Comment (rht) [inline mesa/space.py:144]: Did I make a mistake when I annotated the return type? Seems like this change can be put in a separate commit for clarity.

Comment (Corvince) [reply mesa/space.py:144]: Yes, the return type annotations were wrong. I also just cached this. But despite its name `iter_cell_list_contents` does not return `Iterator[Content]`, but rather `Iterator[Agent]` (since empty cells are not iterated). And this trickles down to all neighbor functions. I'll see what I can do about separating the commits.

Comment (Corvince) [reply mesa/space.py:143]: Good idea to separate the commit. Found several more places were the return type was wrong.

**Comment by rht** [comment]
LGTM

File: mesa/space.py [hunk 5]
```diff
@@ -415,6 +412,25 @@ def exists_empty_cells(self) -> bool:
...
+    def _place_agent(self, pos: Coordinate, agent: Agent) -> None:
+        """ Place the agent at the correct location. """
+        x, y = pos
+        self.grid[x][y] = agent
+        self.empties.discard(pos)
+
+    def _remove_agent(self, pos: Coordinate, agent: Agent) -> None:
+        """ Remove the agent from the given location. """
+        x, y = pos
+        self.grid[x][y] = None
```
Comment (rht) [inline mesa/space.py:430]: Wait, why is this not `self.default_val()`?

Comment (Corvince) [reply mesa/space.py:430]: Hmm it probably should be, but I just copied the function here, so this seems to be an already present bug (and unrelated to this PR)

File: mesa/agent.py (hunk 1 above)
Comment (rht) [reply mesa/agent.py]: Want to know what Agents.jl does for this? They define separate agent struct for each space type, e.g. `ContinuousAgent` https://github.com/JuliaDynamics/Agents.jl/blob/0c660872f983b5a981671bc8f1b5c901c8799a93/src/core/agents.jl#L100

Comment (rht) [reply mesa/agent.py]: This has never been a problem in Python because of duck typing and that methods can have any type, but once we go to explicit typing, we get mess like in Agents.jl.

**Comment by Corvince** [comment]
I have rebased and updated this PR and it is again good to go!

**Comment by rht** [comment]
Closed in favor of #1219. Most notable is that I choose not to implement the generic content type, because the type of the grid content is not interchangeable, unlike `List[...]`, which can be `List[str]`, `List[int]`, and so on.
